lib/dependencies.js

```diff
@@ -0,0 +1,384 @@
+    let oldRootDir = dependencyData.rootDir;
+    newRootDir = newRootDir || oldRootDir;
+    let dependencies = new Dependencies(newRootDir);
+    let files = Object.keys(dependencyData.dependencies);
+    for (let file of files) {
+      let deps = dependencyData.dependencies[file];
+      if (newRootDir) {
+        for (let i = 0; i < deps.length; i++) {
+          let dep = deps[i];
+          if (dep.startsWith(oldRootDir+path.sep)) {
+            deps[i] = dep.replace(oldRootDir, newRootDir);
+          }
+        }
+      }
+      dependencies.setDependencies(file, deps);
+    }
+    /** @type {Map<string, FSTree>} */
+    let fsTrees = new Map();
+    for (let fsTreeData of dependencyData.fsTrees) {
+      /** @type {Array<Entry | HashEntry>} */
+      let entries = [];
+      for (let entry of fsTreeData.entries) {
+        if (entry.type === "stat") {
+          entries.push(new Entry(entry.relativePath, entry.size, entry.mtime, entry.mode));
+        } else {
+          entries.push(new HashEntry(entry.relativePath, entry.hash));
+        }
+      }
+      /** @type {FSTree | FSHashTree}  */
+      let fsTree;
+      /** @type {string} */
+      let treeRoot;
+      if (fsTreeData.fsRoot === oldRootDir) {
+        treeRoot = newRootDir;
+        fsTree = FSHashTree.fromHashEntries(entries, { sortAndExpand: true });
+      } else {
+        treeRoot = fsTreeData.fsRoot;
+        fsTree = FSTree.fromEntries(entries, { sortAndExpand: true });
+      }
+      fsTrees.set(treeRoot, fsTree);
+    }
+    dependencies.seal();
+    dependencies.fsTrees = fsTrees;
+    return dependencies;
+  }
+};
+
+/**
+ * Get an FSTree that uses content hashing information to compare files to
+ * see if they have changed.
+ *
+ * @param fsRoot {string}
+ * @param dependencies {Set<string>}
+ * @return {FSHashTree}
+ */
+function getHashTree(fsRoot, dependencies) {
+  /** @type {Array<HashEntry>} */
+  let entries = [];
+  for (let dependency of dependencies) {
+    let fullPath = path.join(fsRoot, dependency);
+    try {
+      // it would be good if we could cache this and share it with
+      // the read that accompanies `processString()` (if any).
+      let contents = fs.readFileSync(fullPath, "utf8");
+      let hash = md5sum(contents);
+      entries.push(new HashEntry(dependency, hash));
+    } catch(e) {
+      entries.push(new HashEntry(dependency, ""));
+    }
+  }
+  return FSHashTree.fromHashEntries(entries);
+}
+
+/**
+ * Get an FSTree that uses fs.stat information to compare files to see
+ * if they have changed.
+ *
+ * @param fsRoot {string}
+ * @param dependencies {Set<string>}
+ */
+function getStatTree(fsRoot, dependencies) {
+  /** @type {Array<Entry>} */
+  let entries = [];
+  for (let dependency of dependencies) {
+    let fullPath = path.join(fsRoot, dependency);
+    try {
+      // TODO: Share a cache of stat results across all persistent filters.
+      let stats = fs.statSync(fullPath);
```

Comment:
It seems possible for the stat of `fullPath` to have already been found within this invocation of `getStatTree`. Having a local lookup table to see if we have already encountered this fullPath would be a low complexity way of dealing with this. Also, as your comment mentioned future work could expand this to a build wide cache.

Something like:

```js
const seen = { } // fullPath => true | false

// ...
 for (let dependency of dependencies) {
    let fullPath = path.join(fsRoot, dependency);
    if (seen[fullPath]) { continue; }
    seen[fullPath] = true;
    // ...
```